src/frontend/src/optimizer/rule/pull_up_hop_rule.rs

```diff
@@ -0,0 +1,204 @@
+
+        let (new_right,
+            right_time_col,
+            right_window_slide,
+            right_window_size,
+            right_window_offset
+        ) =
+            if let Some(hop) = right.as_logical_hop_window() &&
+            right_input_index_on_condition.iter().all(|&index| hop.output_window_start_col_idx().map_or(true, |v|index!=v) && hop.output_window_end_col_idx().map_or(true, |v|index!=v)) &&
+            join_type != JoinType::LeftAnti && join_type != JoinType::LeftSemi && join_type != JoinType::LeftOuter && join_type != JoinType::FullOuter {
+                let (input,
+                    time_col,
+                    window_slide,
+                    window_size,
+                    window_offset,
+                    _) = hop.clone().into_parts();
+
+                old_i2new_i = old_i2new_i.union(&join.i2r_col_mapping_ignore_join_type().composite(&hop.o2i_col_mapping()).clone_with_offset(new_left.schema().len()));
+                right_output_index = hop.output2internal_col_mapping().try_map_all(right_output_index.into_iter()).unwrap();
+                pull_up_right = true;
+                (input,Some(time_col),Some(window_slide),Some(window_size),Some(window_offset))
+        } else {
+            old_i2new_i = old_i2new_i.union(&join.i2r_col_mapping_ignore_join_type().clone_with_offset(new_left.schema().len()));
+
+            (right,None,None,None,None)
+        };
+
+        if !pull_up_left && !pull_up_right {
+            return None;
+        }
+
+        let new_output_index = {
+            let new_right_output_len =
+                if join_type == JoinType::LeftSemi || join_type == JoinType::LeftAnti {
+                    0
+                } else {
+                    new_right.schema().len()
+                };
+            let new_left_output_len =
+                if join_type == JoinType::RightSemi || join_type == JoinType::RightAnti {
+                    0
+                } else {
+                    new_left.schema().len()
+                };
+            let mut new_output_index =
+                Vec::with_capacity(left_output_index.len() + right_output_index.len());
+            if !pull_up_left {
+                new_output_index.extend(left_output_index.into_iter());
+            } else {
+                new_output_index.extend(left_output_index.into_iter().map(|idx| {
+                    if idx < new_left_output_len {
+                        idx
+                    } else {
+                        idx + new_right_output_len
+                    }
+                }));
+            }
+            if pull_up_right && pull_up_left {
+                new_output_index.extend(right_output_index.into_iter().map(|idx| {
+                    if idx < new_right.schema().len() {
+                        idx + new_left.schema().len()
+                    } else {
+                        idx + new_left.schema().len() + 2
+                    }
+                }));
+            } else {
```

Comment:
Add some assertion? `pull_up_right && ! pull_up_left` or `!pull_up_right && pull_up_left`

Reply:
Above we have the code 
```
    if !pull_up_left && !pull_up_right {
            return None;
    }
```

Obviously it can guarantee the following code is in three case: 
`pull_up_left && pull_up_right `
`!pull_up_left && pull_up_right`
`pull_up_left && !pull_up_right`

So the assertion may isn't needed?🤔